core/ops_metrics.rs

```diff
@@ -1,92 +1,136 @@
 // Copyright 2018-2023 the Deno authors. All rights reserved. MIT license.
 
+use crate::ops::OpCtx;
 use crate::serde::Serialize;
+use crate::OpDecl;
 use crate::OpId;
 use std::cell::Ref;
 use std::cell::RefCell;
 use std::cell::RefMut;
+use std::rc::Rc;
 
-// TODO(@AaronO): split into AggregateMetrics & PerOpMetrics
-#[derive(Clone, Default, Debug, Serialize)]
+/// The type of op metrics event.
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+#[repr(u8)]
+pub enum OpMetricsEvent {
+  /// Entered an op dispatch.
+  Dispatched,
+  /// Left an op synchronously.
+  Completed,
+  /// Left an op asynchronously.
+  CompletedAsync,
+  /// Left an op synchronously with an exception.
+  Error,
+  /// Left an op asynchronously with an exception.
+  ErrorAsync,
+}
+
+/// A callback to receieve an [`OpMetricsEvent`].
+pub type OpMetricsFn = Rc<dyn Fn(&OpCtx, OpMetricsEvent)>;
+
+// TODO(mmastrac): this would be better as a trait
+/// A callback to retrieve an optional [`OpMetricsFn`] for this op.
+pub type OpMetricsFactoryFn =
+  Box<dyn Fn(OpId, usize, &OpDecl) -> Option<OpMetricsFn>>;
+
+#[doc(hidden)]
+pub fn dispatch_metrics_fast(opctx: &OpCtx, metrics: OpMetricsEvent) {
+  // SAFETY: this should only be called from ops where we know the function is Some
+  unsafe { (opctx.metrics_fn.as_ref().unwrap_unchecked())(opctx, metrics) }
+}
+
+#[doc(hidden)]
+pub fn dispatch_metrics_slow(opctx: &OpCtx, metrics: OpMetricsEvent) {
+  // SAFETY: this should only be called from ops where we know the function is Some
+  unsafe { (opctx.metrics_fn.as_ref().unwrap_unchecked())(opctx, metrics) }
+}
+
+#[doc(hidden)]
+pub fn dispatch_metrics_async(opctx: &OpCtx, metrics: OpMetricsEvent) {
+  // SAFETY: this should only be called from ops where we know the function is Some
+  unsafe { (opctx.metrics_fn.as_ref().unwrap_unchecked())(opctx, metrics) }
+}
+
+/// Used for both aggregate and per-op metrics.
+#[derive(Clone, Default, Debug, Serialize, PartialEq, Eq)]
 #[serde(rename_all = "camelCase")]
-pub struct OpMetrics {
-  pub ops_dispatched: u64,
+pub struct OpMetricsSummary {
   pub ops_dispatched_sync: u64,
   pub ops_dispatched_async: u64,
-  // TODO(bartlomieju): this field is never updated
-  pub ops_dispatched_async_unref: u64,
-  pub ops_completed: u64,
-  pub ops_completed_sync: u64,
   pub ops_completed_async: u64,
-  // TODO(bartlomieju): this field is never updated
-  pub ops_completed_async_unref: u64,
-  pub bytes_sent_control: u64,
-  pub bytes_sent_data: u64,
-  pub bytes_received: u64,
 }
 
-// TODO(@AaronO): track errors
-#[derive(Default, Debug)]
-pub struct OpsTracker {
-  ops: RefCell<Vec<OpMetrics>>,
+impl OpMetricsSummary {
+  /// Does this op have outstanding async op dispatches?
+  pub fn has_outstanding_ops(&self) -> bool {
+    self.ops_dispatched_async > self.ops_completed_async
+  }
```

Comment:
Doesn't it need to take ref/unref ops into account?

Reply:
It never did in the past.. those were always zero 🤷